kernel/syscall.c

```diff
@@ -193,20 +193,53 @@ static const char* _format_timespec(
 
 static bool _trace_syscall(long n)
 {
-    // Check if syscall tracing is enabled.
+    /* Check if syscall tracing is enabled. */
     if (__myst_kernel_args.strace_config.trace_syscalls)
     {
         if (__myst_kernel_args.strace_config.filter)
         {
-            // If filtering is enabled, trace only this syscall has been
-            // specified in the filter.
-            return __myst_kernel_args.strace_config.trace[n];
+            /* If filtering is enabled, trace only this syscall if it has been
+               specified in the filter. */
+            if (__myst_kernel_args.strace_config.trace[n] == 0)
+                return false;
         }
-        else
+        if (__myst_kernel_args.strace_config.tid_filter_num > 0)
         {
-            // Trace all syscalls.
-            return true;
+            /* Check if thread filtering is enabled & follow filter */
+            myst_thread_t* self = myst_thread_self();
+            pid_t current_tid = self->tid;
+
+            for (int i = 0; i < __myst_kernel_args.strace_config.tid_filter_num;
+                 i++)
+            {
+                if (current_tid ==
+                    __myst_kernel_args.strace_config.tid_trace[i])
+                    return true;
+            }
+
+            /* tid filter is specified, but no match in filter */
+            return false;
         }
+        if (__myst_kernel_args.strace_config.pid_filter_num > 0)
+        {
+            /* Check if thread filtering is enabled & follow filter */
+            myst_process_t* self = myst_process_self();
+            pid_t current_pid = self->pid;
```

Comment:
```suggestion
            pid_t current_pid = myst_process_self()->pid;
```
It doesn't look like `self` is used anywhere else.

Reply:
I did this on purpose to keep the statements atomic, but I guess the application scope for that is lesser here. Will change.